File: TalentCommon/DataObjects/Base/DBObjectBase.vb
Imports System.IO
Imports System.Text
Imports System.Xml
Imports System.Xml.Serialization
Imports System.Web

#Region "Enumerations"

''' <summary>
''' Provides the status types for the STATUS column in the table tbl_log_header
''' </summary>
Public Enum ActivityStatusEnum
    ''' <summary>
    ''' Start status
    ''' </summary>
    STARTED
    ''' <summary>
    ''' In progress status
    ''' </summary>
    INPROGRESS
    ''' <summary>
    ''' Success
    ''' </summary>
    SUCCESS
    ''' <summary>
    ''' Failed
    ''' </summary>
    FAILED
End Enum

''' <summary>
''' Provides the status types for the STATUS column in the table tbl_data_transfer_status
''' </summary>
Public Enum DataTransferStatusEnum
    ''' <summary>
    ''' Clearing work table (on insert)
    ''' </summary>
    CLEARING_WORK_TABLE
    ''' <summary>
    ''' In progress status (extracting - on update)
    ''' </summary>
    EXTRACTING_FROM_ISERIES
    ''' <summary>
    ''' Updating Tables (on update)
    ''' </summary>
    UPDATING_SQL_TABLES
    ''' <summary>
    ''' Success/Finished (on update)
    ''' </summary>
    FINISHED
    ''' <summary>
    ''' Failed - exception occurred (on update)
    ''' </summary>
    FAILED
End Enum

#End Region

''' <summary>
''' Provides the functionalities which are common across the data objects
''' </summary>
<Serializable()> _
Public MustInherit Class DBObjectBase

#Region "Class Level Fields"
    Private Const CLASSNAME As String = "DBObjectBase"
    ''' <summary>
    ''' Common Cache Key Prefix Start for Data Objects
    ''' </summary>
    Protected Const CACHEKEY_PREFIX As String = "DataObjects_"
    Private _talentLogger As TalentLogging = Nothing
    Private _customObject As Object = Nothing
    Private _agentProfile As DEAgent = Nothing
    Private _resultDataSet As DataSet = Nothing
#End Region

    Sub New()
        _agentProfile = New DEAgent
    End Sub
    ''' <summary>
    ''' Gets or sets the result data set.
    ''' </summary>
    ''' <value>The result data set.</value>
    Public Property ResultDataSet() As DataSet
        Get
            Return _resultDataSet
        End Get
        Set(ByVal value As DataSet)
            _resultDataSet = value
        End Set
    End Property
    Protected ReadOnly Property TalentLogger() As TalentLogging
        Get
            If _talentLogger Is Nothing Then
                _talentLogger = New TalentLogging
            End If
            Return _talentLogger
        End Get
    End Property

    Protected WriteOnly Property CustomObject() As Object
        Set(ByVal value As Object)
            _customObject = value
        End Set
    End Property

    Protected WriteOnly Property AgentProfile() As DEAgent
        Set(ByVal value As DEAgent)
            _agentProfile = value
        End Set
    End Property


#Region "Protected Methods"
    ''' <summary>
    ''' Constructs the parameter using the DESQLParamater instance
    ''' </summary>
    ''' <param name="paramName">Name of the parameter.</param>
    ''' <param name="paramValue">The parameter value.</param>
    ''' <param name="paramType">Type of the parameter. Default SqlDbType.NVarchar</param>
    ''' <param name="paramDirection">The parameter direction. Default ParameterDirection.Input</param>
    ''' <returns>DESQLParameter instance with details</returns>
    Protected Function ConstructParameter(ByVal paramName As String, ByVal paramValue As String, Optional ByVal paramType As SqlDbType = SqlDbType.NVarChar, Optional ByVal paramDirection As ParameterDirection = ParameterDirection.Input) As DESQLParameter
        Dim parameterPiece As New DESQLParameter
        parameterPiece.ParamDirection = paramDirection
        parameterPiece.ParamName = paramName
        parameterPiece.ParamType = paramType
        parameterPiece.ParamValue = paramValue
        Return parameterPiece
    End Function

    ''' <summary>
    ''' Change the given string to upper case
    ''' </summary>
    ''' <param name="stringToUpper">The string to upper.</param>
    ''' <returns>if nothing returns empty string otherwise upper case string</returns>
    Protected Function ToUpper(ByVal stringToUpper As String) As String
        If (Not (stringToUpper = Nothing)) And (stringToUpper.Trim().Length > 0) Then
            stringToUpper = stringToUpper.ToUpper()
        Else
            stringToUpper = String.Empty
        End If
        Return stringToUpper
    End Function

    ''' <summary>
    ''' Construct and return cache key prefix by including common prefix, given class name
    ''' and given method name.
    ''' </summary>
    ''' <param name="className">Name of the class.</param>
    ''' <param name="methodName">Name of the method.</param>
    ''' <returns></returns>
    Protected Function GetCacheKeyPrefix(ByVal className As String, ByVal methodName As String) As String
        Return CACHEKEY_PREFIX & className & methodName
    End Function

    ''' <summary>
    ''' Replaces the single quote to double single quote 
    ''' </summary>
    ''' <param name="stringToReplace">The string to replace.</param>
    ''' <returns>String</returns>
    Protected Function ReplaceSingleQuote(ByVal stringToReplace As String) As String
        If ((stringToReplace <> Nothing) And (stringToReplace.Length > 0)) Then
            stringToReplace = stringToReplace.Replace("'", "''")
        End If
        Return stringToReplace
    End Function

    ''' <summary>
    ''' Converts the Generic List to Xml String based on the given Type
    ''' </summary>
    ''' <typeparam name="T">Type of Object</typeparam>
    ''' <param name="genericListObject">Generic List of the given type.</param>
    ''' <returns>xml string</returns>
    Protected Function GenericListToXmlString(Of T)(ByVal genericListObject As T) As String
        Dim constructedString As String = String.Empty
        Try
            Dim genericSerialiser As New XmlSerializer(GetType(T))
            Dim memoryStreamer As New MemoryStream
            Dim xmlTextWriter As New XmlTextWriter(memoryStreamer, Text.Encoding.Unicode)
            genericSerialiser.Serialize(xmlTextWriter, genericListObject)
            memoryStreamer = xmlTextWriter.BaseStream
            Dim encoding As New UnicodeEncoding()
            constructedString = encoding.GetString(memoryStreamer.ToArray()).Trim()
            genericSerialiser = Nothing
            memoryStreamer = Nothing
        Catch ex As Exception
            Throw
        End Try
        Return constructedString
    End Function

    ''' <summary>
    ''' Remove the given cache key from the cache
    ''' </summary>
    ''' <param name="cacheKey">cache key string</param>
    Protected Sub RemoveKeyFromCache(ByVal cacheKey As String)
        Dim tBase As New TalentBase
        tBase.RemoveItemFromCache(cacheKey)
    End Sub

    Protected Function GetCustomDictionaryEntities(Of T)(ByVal destinationDatabase As DestinationDatabase, ByVal cacheing As Boolean, ByVal cacheTimeMinutes As Integer, ByVal talentSqlAccessDetail As TalentDataAccess, ByVal callingModuleName As String) As T
        Dim customDicObject As T = Nothing
        Dim tempSettingsCacheing As Boolean = False
        Try
            Dim cacheKey As String = talentSqlAccessDetail.Settings.CacheStringExtension
            If talentSqlAccessDetail.Settings.Cacheing AndAlso Talent.Common.TalentThreadSafe.ItemIsInCache(cacheKey) Then
                customDicObject = CType(HttpContext.Current.Cache.Item(cacheKey), T)
            Else
                tempSettingsCacheing = talentSqlAccessDetail.Settings.Cacheing
                talentSqlAccessDetail.Settings.Cacheing = cacheing
                If talentSqlAccessDetail.Settings.Cacheing Then
                    If (talentSqlAccessDetail.Settings.CacheTimeMinutes <= 0 AndAlso talentSqlAccessDetail.Settings.CacheTimeSeconds <= 0) Then
                        talentSqlAccessDetail.Settings.CacheTimeMinutes = cacheTimeMinutes
                    End If
                End If
                Dim outputDataTable As DataTable = Nothing
                'Execute
                Dim err As New ErrorObj
                err = talentSqlAccessDetail.SQLAccessForDictionary(destinationDatabase)
                If (Not (err.HasError)) And (Not (talentSqlAccessDetail.ResultDataSet Is Nothing)) Then
                    outputDataTable = talentSqlAccessDetail.ResultDataSet.Tables(0)
                    If outputDataTable IsNot Nothing AndAlso outputDataTable.Rows.Count > 0 Then
                        PopulateCustomDictionaryEntities(outputDataTable, callingModuleName)
                        customDicObject = _customObject
                        If talentSqlAccessDetail.Settings.Cacheing Then
                            If talentSqlAccessDetail.ModuleNameForCacheDependency.Trim.Length > 0 Then
                                Dim moduleName As String = talentSqlAccessDetail.Settings.ModuleName
                                talentSqlAccessDetail.Settings.ModuleName = talentSqlAccessDetail.ModuleNameForCacheDependency
                                talentSqlAccessDetail.AddItemToCache(cacheKey, customDicObject, talentSqlAccessDetail.Settings)
                                talentSqlAccessDetail.Settings.ModuleName = moduleName
                            Else
                                talentSqlAccessDetail.AddItemToCache(cacheKey, customDicObject, talentSqlAccessDetail.Settings)
                            End If
                        End If
                    ElseIf talentSqlAccessDetail.Settings.Cacheing Then
                        Talent.Common.TalentThreadSafe.RemoveCacheQueueRecord(cacheKey)
                    End If
                End If
            End If
        Catch ex As Exception
            TalentLogger.FrontEndConnectionString = talentSqlAccessDetail.Settings.FrontEndConnectionString
            TalentLogger.Logging(CLASSNAME, "GetCustomDictionaryEntities", "Exception Occured: Calling Module Name is " & callingModuleName, "TCDBOB-001", ex, LogTypeConstants.TCDBDATAOBJECTSLOG, talentSqlAccessDetail.Settings.BusinessUnit, talentSqlAccessDetail.Settings.Partner, talentSqlAccessDetail.Settings.LoginId)
            Throw
        Finally
            talentSqlAccessDetail.Settings.Cacheing = tempSettingsCacheing
            talentSqlAccessDetail = Nothing
        End Try
        Return customDicObject
    End Function
    
    ''' <summary>
    ''' OBSOLETE METHOD
    ''' Use this GetCustomDictionaryEntities(Of T)(ByVal destinationDatabase As DestinationDatabase, ByVal cacheing As Boolean, ByVal cacheTimeMinutes As Integer, ByVal talentSqlAccessDetail As TalentDataAccess, ByVal callingModuleName As String) As T
    ''' </summary>
    ''' <param name="talentSqlAccessDetail"></param>
    ''' <param name="callingModuleName"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Protected Function GetCustomDictionaryEntities(ByVal talentSqlAccessDetail As TalentDataAccess, ByVal callingModuleName As String) As Object
        Dim customDicObject As Object = Nothing
        Try
            Dim cacheKey As String = "GetCustomDictionaryEntities" & talentSqlAccessDetail.Settings.CacheStringExtension
            cacheKey = cacheKey.ToLower
            If talentSqlAccessDetail.Settings.Cacheing AndAlso Talent.Common.TalentThreadSafe.ItemIsInCache(cacheKey) Then
                customDicObject = CType(HttpContext.Current.Cache.Item(cacheKey), Object)
            Else
                Dim outputDataTable As DataTable = Nothing
                'Execute
                Dim err As New ErrorObj
                err = talentSqlAccessDetail.SQLAccess()

                If (Not (err.HasError)) And (Not (talentSqlAccessDetail.ResultDataSet Is Nothing)) Then
                    outputDataTable = talentSqlAccessDetail.ResultDataSet.Tables(0)
                    If outputDataTable IsNot Nothing AndAlso outputDataTable.Rows.Count > 0 Then
                        PopulateCustomDictionaryEntities(outputDataTable, callingModuleName)
                        customDicObject = _customObject
                        If talentSqlAccessDetail.Settings.Cacheing Then
                            talentSqlAccessDetail.AddItemToCache(cacheKey, customDicObject, talentSqlAccessDetail.Settings)
                        End If
                    ElseIf talentSqlAccessDetail.Settings.Cacheing Then
                        Talent.Common.TalentThreadSafe.RemoveCacheQueueRecord(cacheKey)
                    End If
                ElseIf talentSqlAccessDetail.Settings.Cacheing Then
                    Talent.Common.TalentThreadSafe.RemoveCacheQueueRecord(cacheKey)
                End If
            End If
        Catch ex As Exception
            TalentLogger.FrontEndConnectionString = talentSqlAccessDetail.Settings.FrontEndConnectionString
            TalentLogger.Logging(CLASSNAME, "GetCustomDictionaryEntities", "Exception Occured: Calling Module Name is " & callingModuleName, "TCDBOB-001", ex, LogTypeConstants.TCDBDATAOBJECTSLOG, talentSqlAccessDetail.Settings.BusinessUnit, talentSqlAccessDetail.Settings.Partner, talentSqlAccessDetail.Settings.LoginId)
            Throw
        Finally
            talentSqlAccessDetail = Nothing
        End Try

        Return customDicObject
    End Function

    ''' <summary>
    ''' OBSOLETE METHOD
    ''' Use this GetCustomDictionaryEntities(Of T)(ByVal destinationDatabase As DestinationDatabase, ByVal cacheing As Boolean, ByVal cacheTimeMinutes As Integer, ByVal talentSqlAccessDetail As TalentDataAccess, ByVal callingModuleName As String) As T
    ''' </summary>
    ''' <param name="sourceTableForDic"></param>
    ''' <param name="settingsEntity"></param>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Protected Function GetCustomDictionaryEntities(ByVal sourceTableForDic As DataTable, ByVal settingsEntity As DESettings) As Object
        Dim customDicObject As Object = Nothing
        Dim cacheKey As String = "GetCustomDictionaryEntities" & settingsEntity.CacheStringExtension
        cacheKey = cacheKey.ToLower
        If settingsEntity IsNot Nothing AndAlso settingsEntity.Cacheing AndAlso Talent.Common.TalentThreadSafe.ItemIsInCache(cacheKey) Then
            customDicObject = CType(HttpContext.Current.Cache.Item(cacheKey), Object)
        Else
            Try
                If sourceTableForDic IsNot Nothing AndAlso sourceTableForDic.Rows.Count > 0 Then
                    PopulateCustomDictionaryEntities(sourceTableForDic, settingsEntity.ModuleName)
                    customDicObject = _customObject
                    If settingsEntity.Cacheing Then
                        Dim talentSqlAccessDetail As New TalentDataAccess
                        talentSqlAccessDetail.AddItemToCache(cacheKey, customDicObject, settingsEntity)
                        talentSqlAccessDetail = Nothing
                    End If
                ElseIf settingsEntity.Cacheing Then
                    Talent.Common.TalentThreadSafe.RemoveCacheQueueRecord(cacheKey)
                End If
            Catch ex As Exception
                TalentLogger.FrontEndConnectionString = settingsEntity.FrontEndConnectionString
                TalentLogger.Logging(CLASSNAME, "GetCustomDictionaryEntities", "Exception Occured: Calling Module Name is " & settingsEntity.ModuleName, "TCDBOB-002", ex, LogTypeConstants.TCDBDATAOBJECTSLOG, settingsEntity.BusinessUnit, settingsEntity.Partner, settingsEntity.LoginId)
                Throw
            End Try
        End If
        Return customDicObject
    End Function

    Protected Overridable Sub PopulateCustomDictionaryEntities(ByVal dtSourceToPopulate As DataTable, ByVal callingModuleName As String)

    End Sub

    Protected Function TryGetFromCache(Of T)(ByRef isCacheNotFound As Boolean, ByVal cacheing As Boolean, ByVal cacheKey As String) As T
        If cacheing Then
            If Utilities.IsCacheActive Then
                If Not HttpContext.Current.Cache.Item(cacheKey) Is Nothing Then
                    isCacheNotFound = False
                    Return CType(HttpContext.Current.Cache.Item(cacheKey), T)
                End If
            End If
        End If
        isCacheNotFound = True
        Return Nothing
    End Function

#End Region

End Class

File: TalentSystemDefaults/TalentSystemDefaults/DataAccess/Base/DBObjectBase.cs
using IBM.Data.DB2.iSeries;
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using TalentSystemDefaults.DataEntities;
namespace TalentSystemDefaults
{
	#region Enumerations
	/// <summary>
	/// Provides the status types for the STATUS column in the table tbl_log_header
	/// </summary>
	public enum ActivityStatusEnum
	{
		/// <summary>
		/// Start status
		/// </summary>
		STARTED,
		/// <summary>
		/// In progress status
		/// </summary>
		INPROGRESS,
		/// <summary>
		/// Success
		/// </summary>
		SUCCESS,
		/// <summary>
		/// Failed
		/// </summary>
		FAILED
	}
	/// <summary>
	/// Provides the status types for the STATUS column in the table tbl_data_transfer_status
	/// </summary>
	public enum DataTransferStatusEnum
	{
		/// <summary>
		/// Clearing work table (on insert)
		/// </summary>
		CLEARING_WORK_TABLE,
		/// <summary>
		/// In progress status (extracting - on update)
		/// </summary>
		EXTRACTING_FROM_ISERIES,
		/// <summary>
		/// Updating Tables (on update)
		/// </summary>
		UPDATING_SQL_TABLES,
		/// <summary>
		/// Success/Finished (on update)
		/// </summary>
		FINISHED,
		/// <summary>
		/// Failed - exception occurred (on update)
		/// </summary>
		FAILED
	}
	#endregion
	/// <summary>
	/// Provides the functionalities which are common across the data objects
	/// </summary>
	[Serializable()]
	public abstract class DBObjectBase
	{
		#region Class Level Fields
		private const string CLASSNAME = "DBObjectBase";
		private object _customObject = null;
		private DataSet _resultDataSet = null;
		protected string businessUnit = null;
        protected string partnerCode = null;
		protected string variableKey1Value = string.Empty;
		protected string companyCode = string.Empty;
		/// <summary>
		/// Instance of DESettings
		/// </summary>
		protected DESettings settings;
		protected const string ALL_STRING = "*ALL";
		protected const string LANG_ENG = "ENG";
		protected string logSourceClass = string.Empty;
		protected string logSourceMethod = string.Empty;
		protected string logCode = TalentSystemDefaults.GlobalConstants.LOG_ERROR_CODE;
		protected string logFilter1 = string.Empty;
		protected string logFilter2 = string.Empty;
		protected string logFilter3 = string.Empty;
		protected string logFilter4 = string.Empty;
		protected string logFilter5 = string.Empty;
		protected string logContent = string.Empty;
		public enum DBType
		{
			SQLSERVER,
			DB2
		}
		#endregion
		#region Public Properties
		/// <summary>
		/// Gets or sets the result data set.
		/// </summary>
		/// <value>The result data set.</value>
		public DataSet ResultDataSet
		{
			get
			{
				return _resultDataSet;
			}
			set
			{
				_resultDataSet = value;
			}
		}
		/// <summary>
		/// Get database type
		/// </summary>
		/// <value></value>
		/// <returns></returns>
		/// <remarks></remarks>
		private DBType _DBTypeForTable = DBType.SQLSERVER;
		public DBType DBTypeForTable
		{
			get
			{
				return _DBTypeForTable;
			}
			set
			{
				_DBTypeForTable = value;
			}
		}
		#endregion
		#region Public Methods
		/// <summary>
		/// Constructor
		/// </summary>
		/// <param name="settings"></param>
		/// <remarks></remarks>
		public DBObjectBase(DESettings settings)
		{
			this.settings = settings;
			if (settings.VariableKey1 != string.Empty)
			{
				variableKey1Value = settings.VariableKey1;
			}
			if (settings.Company != string.Empty)
			{
				companyCode = settings.Company;
			}
		}
		/// <summary>
		/// Returns list of Current Configurations
		/// </summary>
		/// <param name="configs"></param>
		/// <remarks></remarks>
		public virtual void RetrieveCurrentValues(List<ConfigurationItem> configs)
		{
		}
		/// <summary>
		/// Inserts the mssing configurations
		/// </summary>
		/// <param name="configs"></param>
		/// <param name="currentValues"></param>
		/// <remarks></remarks>
		public virtual void AddMissingValues(System.Collections.Generic.List<ConfigurationItem> configs, Dictionary<string, string> currentValues)
		{
		}
		/// <summary>
		/// Updates the configuration data
		/// </summary>
		/// <param name="configs"></param>
		/// <param name="givenTransaction"></param>
		/// <returns></returns>
		/// <remarks></remarks>
		public virtual int UpdateCurrentValue(System.Collections.Generic.List<ConfigurationItem> configs, SqlTransaction givenTransaction = null)
		{
			int updValue = 0;
			return updValue;
		}
		/// <summary>
		/// Updates the configuration data for DB2
		/// </summary>
		/// <param name="configs"></param>
		/// <param name="givenTransaction"></param>
		/// <returns></returns>
		/// <remarks></remarks>
		public virtual int UpdateCurrentValueDB2(System.Collections.Generic.List<ConfigurationItem> configs, iDB2Transaction givenTransaction = null)
		{
			int updValue = 0;
			return updValue;
		}
		/// <summary>
		/// Inserts Configurations
		/// </summary>
		/// <param name="configs"></param>
		/// <param name="givenTransaction"></param>
		/// <param name="Key"></param>
		/// <returns></returns>
		public virtual int InsertCurrentValue(System.Collections.Generic.List<ConfigurationItem> configs, SqlTransaction givenTransaction, ref string Key)
		{
			int insValue = 0;
			return insValue;
		}
		/// <summary>
		/// Inserts Configurations for DB2
		/// </summary>
		/// <param name="configs"></param>
		/// <param name="givenTransaction"></param>
		/// <param name="Key"></param>
		/// <returns></returns>
		/// <remarks></remarks>
		public virtual int InsertCurrentValueDB2(System.Collections.Generic.List<ConfigurationItem> configs, iDB2Transaction givenTransaction = null)
		{
			return 0;
		}
		/// <summary>
		/// Returns all configurations
		/// </summary>
		/// <param name="businessUnit"></param>
		/// <param name="defaultKeys"></param>
		/// <param name="variableKeys"></param>
		/// <param name="displayName"></param>
		/// <returns></returns>
		/// <remarks></remarks>
		public virtual List<ConfigurationEntity> RetrieveAllValues(string businessUnit, string[] defaultKeys, string[] variableKeys, string displayName = "")
		{
			System.Collections.Generic.List<ConfigurationEntity> lstConfigurationEntity = new System.Collections.Generic.List<ConfigurationEntity>();
			return lstConfigurationEntity;
		}
        /// <summary>
        /// Returns all configurations
        /// </summary>
        /// <param name="businessUnit"></param>
        /// <param name="defaultKeys"></param>
        /// <param name="variableKeys"></param>
        /// <param name="displayName"></param>
        /// <returns></returns>
        /// <remarks></remarks>
        public virtual List<ConfigurationEntity> RetrieveAllValues(string businessUnit, string partner, string[] defaultKeys, string[] variableKeys, string displayName = "")
        {
            System.Collections.Generic.List<ConfigurationEntity> lstConfigurationEntity = new System.Collections.Generic.List<ConfigurationEntity>();
            return lstConfigurationEntity;
        }
		/// <summary>
		/// Returns all iseries configurations
		/// </summary>
		/// <param name="companyCode"></param>
		/// <returns></returns>
		/// <remarks></remarks>
		public virtual List<ConfigurationEntity> RetrieveAlliSeriesValues(string companyCode, string[] defaultKeys, string[] variableKeys, string selectedColumns = "")
		{
			List<ConfigurationEntity> lstConfigurationEntity = new List<ConfigurationEntity>();
			return lstConfigurationEntity;
		}
		/// <summary>
		/// Returns configurations as a list of DEListDetail
		/// </summary>
		/// <param name="dataColumns"></param>
		/// <param name="defaultKeyColumns"></param>
		/// <returns></returns>
		/// <remarks></remarks>
		public virtual List<DEListDetail> RetrieveListValues(System.Collections.Generic.List<DataColumn> dataColumns, System.Collections.Generic.List<KeyColumn> defaultKeyColumns)
		{
			System.Collections.Generic.List<DEListDetail> lstListDetailEntity = new System.Collections.Generic.List<DEListDetail>();
			return lstListDetailEntity;
		}
		/// <summary>
		/// Returns configurations as a list of DEListDetail
		/// </summary>
		/// <returns></returns>
		/// <remarks></remarks>
		public virtual List<DEListDetail> RetrieveListValues()
		{
			System.Collections.Generic.List<DEListDetail> lstListDetailEntity = new System.Collections.Generic.List<DEListDetail>();
			return lstListDetailEntity;
		}
		/// <summary>
		/// Returns configurations as a DataTable
		/// </summary>
		/// <returns></returns>
		/// <remarks></remarks>
		public virtual DataTable RetrieveDTForList()
		{
			DataTable dtForList = new DataTable();
			return dtForList;
		}
		/// <summary>
		/// Returns configurations as a DataTable
		/// </summary>
		/// <param name="dataColumns"></param>
		/// <param name="variableKeyColumns"></param>
		/// <returns></returns>
		/// <remarks></remarks>
		public virtual DataTable RetrieveDTForList(System.Collections.Generic.List<DataColumn> dataColumns, System.Collections.Generic.List<KeyColumn> variableKeyColumns)
		{
			DataTable dtForList = new DataTable();
			return dtForList;
		}
		/// <summary>
		/// Deletes configurations
		/// </summary>
		/// <param name="givenTransaction"></param>
		/// <returns></returns>
		/// <remarks></remarks>
		public virtual int DeleteCurrentValues(SqlTransaction givenTransaction = null)
		{
			int dltValue = 0;
			return dltValue;
		}
		/// <summary>
		/// Deletes configurations in DB2
		/// </summary>
		/// <param name="givenTransaction"></param>
		/// <returns></returns>
		public virtual int DeleteCurrentValuesDB2(iDB2Transaction givenTransaction = null)
		{
			return 0;
		}
		#endregion
		#region Protected Methods
		/// <summary>
		/// Constructs the parameter using the DESQLParamater instance
		/// </summary>
		/// <param name="paramName">Name of the parameter.</param>
		/// <param name="paramValue">The parameter value.</param>
		/// <param name="paramType">Type of the parameter. Default SqlDbType.NVarchar</param>
		/// <param name="paramDirection">The parameter direction. Default ParameterDirection.Input</param>
		/// <returns>DESQLParameter instance with details</returns>
		protected DESQLParameter ConstructParameter(string paramName, object paramValue, SqlDbType paramType = SqlDbType.NVarChar, ParameterDirection paramDirection = ParameterDirection.Input)
		{
			DESQLParameter parameterPiece = new DESQLParameter();
			parameterPiece.ParamDirection = paramDirection;
			parameterPiece.ParamName = paramName;
			parameterPiece.ParamType = paramType;
			parameterPiece.ParamValue = paramValue;
			return parameterPiece;
		}
		/// <summary>
		/// Constructs the parameter using the DEDB2Parameter instance
		/// </summary>
		/// <param name="paramName"></param>
		/// <param name="paramValue"></param>
		/// <param name="paramType"></param>
		/// <param name="paramDirection"></param>
		/// <returns></returns>
		/// <remarks></remarks>
		protected DEDB2Parameter ConstructDB2Parameter(string paramName, string paramValue, iDB2DbType paramType = iDB2DbType.iDB2Char, ParameterDirection paramDirection = ParameterDirection.Input)
		{
			DEDB2Parameter parameterPiece = new DEDB2Parameter();
			parameterPiece.ParamDirection = paramDirection;
			parameterPiece.ParamName = paramName;
			parameterPiece.ParamType = paramType;
			parameterPiece.ParamValue = paramValue;
			return parameterPiece;
		}
		/// <summary>
		/// Constructs the parameter using the DEDB2Paramater instance
		/// </summary>
		/// <param name="paramName">Name of the parameter.</param>
		/// <param name="paramValue">The parameter value.</param>
		/// <param name="paramType">Type of the parameter. Default iDB2DbType.Char</param>
		/// <param name="paramDirection">The parameter direction. Default ParameterDirection.Input</param>
		/// <returns>DEDB2Parameter instance with details</returns>
		protected DEDB2Parameter ConstructDB2Parameter(string paramName, string paramValue, int paramSize, iDB2DbType paramType = iDB2DbType.iDB2Char, ParameterDirection paramDirection = ParameterDirection.Input)
		{
			DEDB2Parameter parameterPiece = new DEDB2Parameter();
			parameterPiece.ParamDirection = paramDirection;
			parameterPiece.ParamName = paramName;
			parameterPiece.ParamType = paramType;
			parameterPiece.ParamValue = paramValue;
			parameterPiece.ParamLength = paramSize;
			return parameterPiece;
		}
		/// <summary>
		/// Change the given string to upper case
		/// </summary>
		/// <param name="stringToUpper">The string to upper.</param>
		/// <returns>if nothing returns empty string otherwise upper case string</returns>
		protected string ToUpper(string stringToUpper)
		{
			if ((!(stringToUpper == null)) && (stringToUpper.Trim().Length > 0))
			{
				stringToUpper = stringToUpper.ToUpper();
			}
			else
			{
				stringToUpper = string.Empty;
			}
			return stringToUpper;
		}
		protected int DB2Access(TalentDB2Access talentDB2AccessDetail, iDB2Transaction givenTransaction)
		{
			int affectedRows = 0;
			if (settings.EnableDB2AccessQueue)
			{
				settings.DB2AccessQueue.Add(talentDB2AccessDetail);
			}
			else
			{
				ErrorObj err = new ErrorObj();
				if (givenTransaction == null)
				{
					err = talentDB2AccessDetail.DB2Access(DestinationDatabase.TALENTTKT);
				}
				else
				{
					err = talentDB2AccessDetail.DB2Access(DestinationDatabase.TALENTTKT, givenTransaction);
				}
				if ((!(err.HasError)) && (!(talentDB2AccessDetail.ResultDataSet == null)))
				{
					affectedRows = System.Convert.ToInt32(talentDB2AccessDetail.ResultDataSet.Tables[0].Rows[0][0]);
				}
			}
			return affectedRows;
		}
		#endregion
	}
}
